File: app/src/main/java/ardjomand/leonardo/nutrimeal/meals/MealsFragment.java
package ardjomand.leonardo.nutrimeal.meals;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

import ardjomand.leonardo.nutrimeal.BuildConfig;
import ardjomand.leonardo.nutrimeal.R;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;

/**
 * A fragment representing a list of Items.
 * <p/>
 * Activities containing this fragment MUST implement the {@link OnMealFragmentInteractionListener}
 * interface.
 */
public class MealsFragment extends Fragment implements
        MealsContract.View,
        MealAdapter.OnMealAdapterInteractionListener {

    private static final String ARG_COLUMN_COUNT = "column-count";

    //region Views
    @BindView(R.id.button)
    Button button;
    @BindView(R.id.list)
    RecyclerView recyclerView;
    @BindView(R.id.meals_add_fab)
    FloatingActionButton addMeal;
    //endregion

    //region Member variables
    private int mColumnCount = 1;
    private OnMealFragmentInteractionListener mListener;
    private Unbinder unbinder;
    private MealsPresenter presenter;
    private MealAdapter adapter;
    //endregion

    //region Constructors
    /**
     * Mandatory empty constructor for the fragment manager to instantiate the
     * fragment (e.g. upon screen orientation changes).
     */
    public MealsFragment() {
    }

    @SuppressWarnings("unused")
    public static MealsFragment newInstance(int columnCount) {
        MealsFragment fragment = new MealsFragment();
        Bundle args = new Bundle();
        args.putInt(ARG_COLUMN_COUNT, columnCount);
        fragment.setArguments(args);
        return fragment;
    }
    //endregion

    //region Lifecycle
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);

        if (getArguments() != null) {
            mColumnCount = getArguments().getInt(ARG_COLUMN_COUNT);
        }

        presenter = new MealsPresenter(this);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_meals, container, false);
        unbinder = ButterKnife.bind(this, view);

        setTitle();

        // Set layout manager
        Context context = view.getContext();
        if (mColumnCount <= 1) {
            recyclerView.setLayoutManager(new LinearLayoutManager(context));
        } else {
            recyclerView.setLayoutManager(new GridLayoutManager(context, mColumnCount));
        }

        // Set adapter
        adapter = new MealAdapter(new ArrayList<Meal>(), this, getContext());
        recyclerView.setAdapter(adapter);

        // Set decoration
        RecyclerView.ItemDecoration itemDecoration = new
                DividerItemDecoration(recyclerView.getContext(), DividerItemDecoration.VERTICAL);
        recyclerView.addItemDecoration(itemDecoration);

        return view;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnMealFragmentInteractionListener) {
            mListener = (OnMealFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnOrdersFragmentInteractionListener");
        }
    }

    @Override
    public void onStart() {
        super.onStart();
        adapter.clearData();
        presenter.subscribeToMealsUpdates();
    }

    @Override
    public void onStop() {
        super.onStop();
        presenter.unsubscribeFromMealsUpdates();
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenter.setView(null);
    }
    //endregion

    //region Menu
    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_meals_fragment, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_item_orders:
                mListener.onGoToOrdersClicked();
                break;
        }
        return super.onOptionsItemSelected(item);
    }
    //endregion

    //region Presenter callbacks
    @Override
    public void showMeals(List<Meal> meals) {
        adapter.replaceData(meals);
    }

    @Override
    public void addMeal(Meal meal) {
        adapter.addData(meal);
    }

    @Override
    public void updateMeal(Meal meal) {
        adapter.updateData(meal);
    }

    @Override
    public void showEmptyMeals() {
        Toast.makeText(getActivity(), "No meals available", Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showError() {
        Toast.makeText(getActivity(), "Error", Toast.LENGTH_SHORT).show();
    }

    @Override
    public void goToEditMeal(String key) {
        mListener.onEditMealClicked(key);
    }

    @Override
    public void goToEditCartMealQuantity(String key) {
        mListener.onEditCartMealQuantity(key);
    }
//endregion

    private void setTitle() {
        if (getActivity() instanceof AppCompatActivity) {
            ActionBar supportActionBar = ((AppCompatActivity) getActivity()).getSupportActionBar();
            if (supportActionBar != null) {
                supportActionBar.setTitle(R.string.meals_title);
            }
        }
    }

    @OnClick({R.id.button, R.id.meals_add_fab})
    public void onViewClicked(View view) {
        switch (view.getId()) {
            case R.id.button:
                mListener.onGoToCartClicked();
                break;

            case R.id.meals_add_fab:
                presenter.createNewMeal();
                break;
        }
    }

    @Override
    public void onMealClicked(Meal meal) {
        if (BuildConfig.FLAVOR.equals("company")) {
            presenter.editMeal(meal.getKey());
        } else {
            // TODO add meal to cart only if item quantity in cart is zero
            presenter.addMealToCart(meal);
            presenter.editCartMealQuantity(meal.getKey());
            // TODO only show toast if meal was successfully added to cart
//            Toast.makeText(getContext(), meal.getName() + " added to cart", Toast.LENGTH_SHORT).show();
        }
    }

    public interface OnMealFragmentInteractionListener {
        void onEditMealClicked(String key);
        void onEditCartMealQuantity(String key);
        void onGoToCartClicked();
        void onGoToOrdersClicked();
    }
}

File: app/src/main/java/ardjomand/leonardo/nutrimeal/meals/MealsFragment.kt
package ardjomand.leonardo.nutrimeal.meals

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.LinearLayoutManager
import android.view.*
import android.widget.Toast
import ardjomand.leonardo.nutrimeal.BuildConfig
import ardjomand.leonardo.nutrimeal.R
import ardjomand.leonardo.nutrimeal.data.pojos.Meal
import kotlinx.android.synthetic.company.fragment_meals.*
import java.util.*

class MealsFragment : Fragment(),
        View.OnClickListener,
        MealsContract.View<Meal>,
        MealAdapter.OnMealAdapterInteractionListener {

    //region Member variables
    private var columnCount = 1
    private var listener: OnMealFragmentInteractionListener? = null
    private lateinit var presenter: MealsContract.Presenter
    private lateinit var adapter: MealAdapter
    //endregion

    //region Lifecycle
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setHasOptionsMenu(true)

        arguments?.let {
            columnCount = it.getInt(ARG_COLUMN_COUNT)
        }

        presenter = MealsPresenter(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_meals, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        setTitle()

        val context = view.context

        // Set layout manager
        list.layoutManager = when {
            columnCount <= 1 -> LinearLayoutManager(context)
            else -> GridLayoutManager(context, columnCount)
        }

        // Set adapter
        adapter = MealAdapter(ArrayList(), this, context)
        list.adapter = adapter

        // Set decoration
        val itemDecoration = DividerItemDecoration(list.context, DividerItemDecoration.VERTICAL)
        list.addItemDecoration(itemDecoration)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is OnMealFragmentInteractionListener) {
            listener = context
        } else {
            throw RuntimeException(context.toString() + " must implement OnOrdersFragmentInteractionListener")
        }
    }

    override fun onStart() {
        super.onStart()
        adapter.clearData()
        presenter.subscribe()
    }

    override fun onStop() {
        super.onStop()
        presenter.unsubscribe()
    }

    override fun onDetach() {
        super.onDetach()
        listener = null
    }

    override fun onDestroy() {
        super.onDestroy()
        presenter.setView(null)
    }
    //endregion

    //region Menu
    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater) {
        inflater.inflate(R.menu.menu_meals_fragment, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_item_orders -> listener?.onGoToOrdersClicked()
        }
        return super.onOptionsItemSelected(item)
    }
    //endregion

    //region Presenter callbacks
    override fun addItem(meal: Meal) {
        adapter.addData(meal)
    }

    override fun updateItem(meal: Meal) {
        adapter.updateData(meal)
    }

    override fun showEmptyItems() {
        Toast.makeText(activity, R.string.error_no_items_available, Toast.LENGTH_SHORT).show()
    }

    override fun showError() {
        Toast.makeText(activity, R.string.error_items_could_not_be_downloaded, Toast.LENGTH_SHORT).show()
    }

    override fun goToEditMeal(key: String) {
        listener?.onEditMealClicked(key)
    }

    override fun goToEditCartMealQuantity(key: String) {
        listener?.onEditCartMealQuantity(key)
    }
    //endregion

    // TODO move the responsibility of setting title to the parent activity
    // this will make it easier for the fragments to be added instead of replaced
    private fun setTitle() {
        if (activity is AppCompatActivity) {
            val supportActionBar = (activity as AppCompatActivity).supportActionBar
            supportActionBar?.apply {
                setTitle(R.string.meals_title)
                setDisplayHomeAsUpEnabled(false)
            }
        }
    }

    override fun onClick(view: View?) {
        when (view?.id) {
            R.id.button -> listener?.onGoToCartClicked()

            R.id.meals_add_fab -> presenter.createNewMeal()
        }
    }

    override fun onMealClicked(meal: Meal) {
        if (BuildConfig.FLAVOR == "company") {
            presenter.editMeal(meal.key)
        } else {
            // TODO add meal to cart only if item quantity in cart is zero
            presenter.addMealToCart(meal)
            presenter.editCartMealQuantity(meal.key)
            // TODO only show toast if meal was successfully added to cart
            //            Toast.makeText(getContext(), meal.getName() + " added to cart", Toast.LENGTH_SHORT).show();
        }
    }

    interface OnMealFragmentInteractionListener {
        fun onEditMealClicked(key: String)
        fun onEditCartMealQuantity(key: String)
        fun onGoToCartClicked()
        fun onGoToOrdersClicked()
    }

    companion object {

        private const val ARG_COLUMN_COUNT = "column-count"

        fun newInstance(columnCount: Int) =
                MealsFragment().apply {
                    arguments = Bundle().apply {
                        putInt(ARG_COLUMN_COUNT, columnCount)
                    }
                }
    }
}
